src/wasm/wasm-validator.cpp

```diff
@@ -2749,6 +2749,17 @@ void FunctionValidator::visitRefCast(RefCast* curr) {
         curr->ref->type.isRef(), curr, "ref.cast ref must have ref type")) {
     return;
   }
+  // If the cast is unreachable but not the ref (we ruled out the former
+  // earlier), then the cast is unreachable because the cast type had no
+  // common supertype with the ref, which is invalid. This is the same as the
+  // check below us, but we must do it first (as getHeapType fails otherwise).
+  if (!shouldBeUnequal(
```

Comment:
We could tighten this up even more by checking that the result type is a reference type. So if it somehow was set to i32 or something, we would still avoid the assertion failure.

Reply:
Sounds good, I added that now as an extra test (it has a more generic error message, and I didn't want to lose the specific explanation for the supertype issue).